### ISSUE TITLE
과적합 방지 관련

### ISSUE THREAD

**Issue by YJin33** [KnittingGirls/AI#12]
weight decay 적용
- 원래 하려고 했으나 오타로 제대로 되지 않고
- weight decay와 lr_scheduler를 동시에 적용하면 제대로 되지 않는 것 같음.
- 둘 중 한 가지만 적용하는 게 나을 것으로 보임!

현재 DeepLabV3_ujin 에서는 epoch 20, learning scheduler 변경, loss=focal loss로 적용된 상태.
자세한 것은 notion 링크 참고하기

**Comment by YJin33** [comment]
값 조정해서 weight decay와 lr_scheduler 모두 적용해둠. 아직 더 조정 필요한 상태
![image](https://github.com/user-attachments/assets/6b0ac555-ae84-46de-9751-5415fb3bceae)
![image](https://github.com/user-attachments/assets/cbb44bd0-ab58-492d-a67b-00c6a8f55a4b)

**Comment by YJin33** [comment]
optimizer = torch.optim.Adam([
    dict(params=model.parameters(), lr=0.0001, weight_decay=1e-5),
])
lr_scheduler = CosineAnnealingWarmUpRestarts(optimizer, T_0=10, T_mult=1, eta_max=0.001,  T_up=2, gamma=0.5)

lr scheduler는 자체 정의된 것, 다른 사람이 정의해둔 것을 활용함

**Comment by YJin33** [comment]
optimizer = torch.optim.Adam([
    dict(params=model.parameters(), lr=0.0001, weight_decay=1e-5),
])
lr_scheduler = CosineAnnealingWarmUpRestarts(optimizer, T_0=10, T_mult=1, eta_max=0.01,  T_up=2, gamma=0.3)

![image](https://github.com/user-attachments/assets/293a6176-8a9a-4826-aeb5-e40b9ee3e467)
![image](https://github.com/user-attachments/assets/138eec6c-e86f-4387-ace1-f244240ea2b3)

**Comment by YJin33** [comment]
Early stopping 오픈 소스 활용해서 구현해둠.
epoch 9에서 stop하길래 epoch 26까지는 무조건 작동하도록 에포크 20개를 다 돌고 나서부터 early stopping 적용하도록 구현함.
![image](https://github.com/user-attachments/assets/45cc67ae-8ae3-4562-860d-4fc4729cd744)
![image](https://github.com/user-attachments/assets/62354122-d387-4011-b1de-6a7d751fdf54)
 

성능 지표 여러개로 늘려서 작동하는 것 확인함. df 있는 셀 확인해보면, accuracy는 꽤 높지만 다른 요소가 좋지 않음을 알 수 있음. 
-> 이부분을 처리해야 할 것 같음.

데이터가 상당히 적은 클래스(파란색)이 과도하게 나타나는 경향이 보임 - focal loss의 영향일 수도 있다고 생각. 
일단은 distance 계산하는 부분에서 클래스간 거리를 모두 계산해서 비교하는 식으 수정해볼 예정!
